`validate_no_conflicts_unreliably/3`

nathanl opened the pull request:
## The Feature

I've [advocated](https://elixirforum.com/t/a-case-for-validating-uniqueness/3637) that Ecto support *validating* uniqueness, despite the fact that a constraint must also be used to prevent race conditions. My reasons are:

- If someone has taken the username I want, they *probably* took it yesterday or last year, not a few milliseconds ago
- In a server-rendered app, it's annoying to fix the "required" and "format" validations, etc, only to submit and get *another* error saying "username is taken" (and possibly another for each constraint we have). It's nice if we can tell the user "email is required and username is already taken" at the same time (except in the case of race conditions). This is what Rails does normally (although it doesn't handle the constraint violation gracefully, if it happens).

@josevalim [seemed OK with the concept](https://elixirforum.com/t/a-case-for-validating-uniqueness/3637/13) as long as it was named `unsafe_validate_unique` or something similar, to warn people that they still need a constraint for guaranteed uniqueness.

## The Implementation

This code allows calling:

- `unsafe_validate_unique(changeset, [:username], MyApp.Repo)`
- `unsafe_validate_unique(changeset, [:city, :state], MyApp.Repo)`
- `unsafe_validate_unique(changeset, [:city, :state], MyApp.Repo, :city, "not unique in state")`

I *think* this implementation is roughly OK, but I'm not sure the best way to test it. I think I should persist one changeset, then validate some other changesets given the existence of the first record in the database.

But no existing validation tests use a repo.

Suggestions on where / how to test this without ruining the existing test suite? :) Outright rejection of the feature?

josevalim commented on lib/ecto/changeset.ex:
```diff
@@ -1349,6 +1349,43 @@ defmodule Ecto.Changeset do
 
+  import Ecto.Query # probably not the right place to do this?
+  # @spec  ??
+  def unsafe_validate_unique(changeset, field_names, repo, error_field \\ :base, error_message \\ "not unique") do
```
Ecto doesn't have a `:base` field. :)

josevalim replied:
API-wise, I wonder if it would be better to allow this:

    unsafe_validate_unique(changeset, :username, repo, "not unique")

If you want other considerations, then you can do this:

    unsafe_validate_unique(changeset, :username, &look_with_scope/1, "not unique")

where

     defp look_with_scope(query) do
       Repo.all from q in query, where: ...
     end

This way we don't need to support scoped queries... the user will rather do it directly. Passing a repo becomes just syntax sugar.

OvermindDL1 replied:
We also need some way to pass in a custom where clause as well.  For example, my email field has a unique index that uses a SQL function that is lowercasing it and trimming it and such, and since this is not using the uniqueness setup from the database itself then we need to be able to specify it manually here (still seems easily desyncable...).

Also, this really needs some marker saying `I AM SLOW BECAUSE I DO A DB QUERY`, unsure `unsafe` is quite accurate enough, but it is indeed unsafe regardless.  The repo variable may not be enough to signify this?

michalmuskala commented:
To be honest, I'm not sure this feature really belongs in changeset. 
Currently, changesets are blissfully unaware of repositories or queries (besides an opaque field called `repo`), and this changes it.

If we ever want to split the changeset part away from the database part of ecto (and this is something, I'd like to do), this will make it much, much harder.

This makes it a show-stopper for me. I understand the usefulness of the feature, but it will introduce a huge maintanance burden in the future by tying several parts together, that are not tied today.

ericmj commented:
From @OvermindDL1:
> We also need some way to pass in a custom where clause as well. For example, my email field has a unique index that uses a SQL function that is lowercasing it and trimming it and such, and since this is not using the uniqueness setup from the database itself then we need to be able to specify it manually here (still seems easily desyncable...).

From @michalmuskala:
> If we ever want to split the changeset part away from the database part of ecto (and this is something, I'd like to do), this will make it much, much harder.

@josevalim's proposal of using a function instead of passing in the repo solves both of these concerns.

nathanl commented:
I made a few changes, based on this feedback (and some further thought).

- Updated to always take a function that queries for conflicting records
- Renamed to `validate_no_conflicts_unreliably`, because:
  - I like having it start with `validate_`, like the other validation functions
  - `unreliably` seems more to the point than `unsafe`
  - Since the user will provide the function, it's more about `conflicts` than uniqueness; this can just as easily be used to check for overlapping rentals that would violate an exclusion constraint

The upside of this change is total decoupling from the database and increased flexibility.

The downside of this change is that the function doesn't really do much for you; all the real work is done by the function passed in.

I'm OK with that, though. I just want to make this kind of technique more visible and normal, so that users get a better experience. Having this listed and explained in the docs will accomplish that.

nathanl commented:
I had one other idea - handling the "don't consider updates to the same record to be a conflict" check into this function, so the user-defined functions don't all have to do that.

nathanl commented on lib/ecto/changeset.ex:
```diff
@@ -1349,6 +1349,67 @@ defmodule Ecto.Changeset do
+    conflicts = changeset
+    |> conflict_finder_func.()
+    |> Enum.reject(fn (conflict) ->
+      conflict.id == get_field(changeset, :id)
```
Is it a problem to assume there's a unique `:id` field? If so, should this "a record doesn't conflict with itself" check also be handled by the `conflict_finder_func`?

josevalim replied:
Yes. `struct.__schema__(:primary_key)` will return all primary keys. That would be the preferred mechanism.

nathanl replied:
K. Will fix.

nathanl commented:
Updated so that the validation checks for updates - if the user's function says "looks like the changeset would conflict with this record", but the changeset would actually update that record (which we can tell because they are for the same schema and the same primary key), the validation doesn't consider that a conflict.

nathanl commented:
Any more concerns or suggestions?

shdblowers commented:
In its current form I don't feel like it adds much more functionality on top of something that's more generic like `Ecto.Changeset.validate_change`.

Also, as an end-user I'd feel very wary about using a function that had `unreliably` in the name.

johnnyshields commented:
:+1: for this feature, the `validate_unique` should be fine as that's what most other frameworks are calling it...

nathanl commented:
> 👍 for this feature, the `validate_unique` should be fine as that's what most other frameworks are calling it...

Yep, that's the common name. However, "not unique" is just variety of possible conflicts with existing data - see the "overlap" example I gave in my previous comment. 

In general, PostgreSQL exclusion constraints can check equality, inequality, and overlap between values in two rows. I spent a bunch of time reading about this and [wrote it up on my blog](http://nathanmlong.com/2016/01/protect-your-data-with-postgresql-constraints/) a while back. The example using inequality that I gave was this:

> To explain, I'll borrow an example from a presentation by Jeff Davis, who implemented exclusion constraints for PostgreSQL 9.0. Jeff said that in a zoo management system, you could express the idea "all animals assigned to the same enclosure must be the same species" by saying "don't allow new assignments rows where the enclosure_id is equal (=) to another assignment's AND the species_id is unequal (<>) to the other assignment's".

nathanl commented:
> `validate_no_conflicts_unreliably` is meant to serve as a "pre-check" for constraints, and the given function may need to check multiple fields.

A possible alternative name would be something like `pre_validate_constraint`, but that would only make sense if we supply the constraint name to the function and infer the errors from the existing constraint.

nathanl commented:
@michalmuskala What do you think of the current implementation?

nathanl commented on lib/ecto/changeset.ex:
```diff
@@ -1349,6 +1349,99 @@ defmodule Ecto.Changeset do
+    changeset.data.__struct__.__schema__(:source)
+  end
+
+  defp source(record) do
```
Maybe the versions of `source/1` and `primary_key/1` that take a record should belong to a different module?

nathanl commented on mix.lock:
```diff
@@ -1,12 +1,12 @@
-  "poolboy": {:hex, :poolboy, "1.5.1", "6b46163901cfd0a1b43d692657ed9d7e599853b3b21b95ae5ae0a777cf9b6ca8", [], [], "hexpm"},
-  "postgrex": {:hex, :postgrex, "0.13.2", "2b88168fc6a5456a27bfb54ccf0ba4025d274841a7a3af5e5deb1b755d95154e", [:mix], [{:connection, "~> 1.0", [hex: :connection, repo: "hexpm", optional: false]}, {:db_connection, "~> 1.1", [hex: :db_connection, repo: "hexpm", optional: false]}, {:decimal, "~> 1.0", [hex: :decimal, repo: "hexpm", optional: false]}], "hexpm"},
-  "sbroker": {:hex, :sbroker, "1.0.0", "28ff1b5e58887c5098539f236307b36fe1d3edaa2acff9d6a3d17c2dcafebbd0", [:rebar3], [], "hexpm"}}
+  "connection": {:hex, :connection, "1.0.4", "a1cae72211f0eef17705aaededacac3eb30e6625b04a6117c1b2db6ace7d5976", [:mix], []},
```
These changes are unrelated to the PR, but I couldn't run the tests without updating the dependencies.

nathanl commented:
@josevalim Any feedback? (Sorry to bug you.)

josevalim commented:
It is fine to bug but unfortunately I do not have much cycles right now to put on Ecto. We have ElixirConf EU and upcoming Erlang versions so all of the work is going towards that for now. :) Hopefully by July we can work on getting Ecto 2.2 out.

mbuhot commented:
This feels like a special case of `validate_change` or just a custom validation with `add_error`:

Validating that a post title doesn't exist:

```elixir
validate_change(changeset, :title, &validate_title_unused/1)

defp validate_title_unused(:title, title) do
  case Repo.get_by(Post, title: title) do
    nil -> []
    _ -> [title: "already in use"]
  end
end
```

Validating no overlapping rental for given property is complex enough to use a custom validation:

```elixir
def new_booking_changeset(params) do
  %Booking{}
  |> cast(params, [:start_date, :end_date, :property_id])
  |> validate_required([:start_date, :end_date, :property_id])
  |> validate_no_overlapping()
end

defp validate_no_overlapping(changeset) do
  start_date = Changeset.get_field(:start_date)
  end_date = Changeset.get_field(changeset, :end_date)
  property_id = Changeset.get_field(changset, :property_id)
  query = find_rentals_by_date(property_id, start_date, end_date) |> limit(1)
  case  Repo.one(query) do
    nil -> changeset
    _ -> Changeset.add_error(changeset, :start_date, "property is not available")
end
```

It also doesn't feel like a very efficient approach to query the database for a list of _potential_ conflicts, then have Ecto check if it would update those records in the current changeset.
Conflicts can usually be detected with an `exists` subquery that doesn't need to pull any data back.

nathanl commented:
> It also doesn't feel like a very efficient approach to query the database for a list of potential conflicts, then have Ecto check if it would update those records in the current changeset.
> Conflicts can usually be detected with an `exists` subquery that doesn't need to pull any data back.

Good point - checking `exists` would be much more efficient. OTOH, it would mean that the user function would have to add "and it isn't the id (if any) of the current changeset" to its query, and that this function would be doing little more than checking a boolean value returned by the user function.

I'll play around with it. My main goal is just that "a preliminary check for conflicts at validation time" becomes an easy and obvious thing to do with Ecto.

nathanl commented:
@mbuhot 

I think `validate_change` is insufficient for checking conflicts even on a single field, because it doesn't pass the changeset into the validator function. For example, your `validate_title_unused` can see if the title exists in the database, but that doesn't necessarily mean there's a conflict; the current changeset could be an update to the post with that title.

You might think "well, only check `validate_title_unused` when inserting", but that doesn't work either. You need to validate on update also, otherwise you'd miss cases where an existing post is being updated with a title another post already had. We'd have to see whether the changeset and existing record have the same id and schema to know if this is an update to the existing record.

Your `validate_no_overlapping` function also should check whether the changeset and existing record (if any) are the same id & schema.

It's kind of a lot to think about, isn't it? This is why I'd like to make it easier. But if (as you point out) the most efficient thing for the user-supplied function to do is to run an `exists` query, that reduces `validate_no_conflicts_unreliably/3` to checking a boolean and setting an error, which means it doesn't add much real value, other than being a placeholder in the docs for this concept.

nathanl commented:
I think that @michalmuskala has argued convincingly that we don't want a built-in validation that queries a database.

And I think @mbuhot has argued convincingly that we don't want a built-in validation that looks through potential conflicts in Elixir when a user-defined `exists` query would be much more efficient.

That leaves me thinking that a custom validation is the way that Ecto users should handle this.

```elixir
  def validate_no_conflicting_names(changeset = %Ecto.Changeset{changes: %{name: name}}) when not is_nil(name) do
    case name do
      nil -> false
      _ ->
        dups_query = from e in Event, where: e.name == ^name
        # For updates, don't flag event as a dup of itself
        id = get_field(changeset, :id)
        dups_query = if is_nil(id) do
          dups_query
        else
          from e in dups_query, where: e.id != ^id
        end

        exists_query = from q in dups_query, select: true, limit: 1
        case Repo.one(exists_query) do
          true -> add_error(
            changeset, :name, "has already been taken", [validation: :validate_no_conflicting_names]
          )
          nil  -> changeset
        end
    end
  end

  # If event has no name or a nil name, it isn't a conflict
  def validate_no_conflicting_names(changeset), do: changeset
```

So there's probably no reason to add a new function to Ecto for this case.

However, right now, the docs say that "validations can be executed without a need to interact with the database", which I think discourages people from checking for conflicts in the validation phase, even though:

- 99.99% of the time, if your desired username is taken, it was taken before your input was validated
- It's much nicer to fix all errors in one pass than it is to fix validation errors, then fix a constraint error, then fix another constraint error...

Should I try a separate PR where I simply update the docs to suggest that you *might* want to check for conflicting data in the validation phase?

johnnyshields commented:
@nathanl there are many use cases to doing a simple uniqueness validation before/without saving--client side validations for example. It's something that we'd definitely want in the toolkit; we shouldn't have to write 20 lines of boilerplate code each time it's needed.

mbuhot commented:
@johnnyshields: Maybe this uniqueness helper could be exposed in a separate hex package? That relaxes the requirement that the validation function doesn't access the Repo.

josevalim commented:
@nathanl there is also a chance that this PR suffered the problem of being too generic. We made it generic enough to the point that using it gives us very little compared to rolling our own. Do you have the `unsafe_validate_uniquess` PR around by any chance so we can compare the current PR with a specialized solution?

kelvinst commented:
Well, looking really quick to the discussion I thought (naively) in a different way of solving the problem: running the unique constraints in validation time.

Doing a little bit of search I realized that there is no way of executing a constraint without executing the `UPDATE` or `INSERT` on the database itself. Shame, but there is a reason to it, I know. 

But seriously, why not? We could do something like:

- Run validations
  - If `changeset.valid?` just do it and let the constraints fail!!
  - If not and we have `unique_constraint/3` (or other constraint validations) applied to the `changeset`:
    - Create a fake transaction
    - Do it and let the constraints fail if invalid
    - Rolback the transaction

I know the solutions looks "ugly", I'm just trying to open my mind.

kelvinst commented:
Bad things about it:

- It's "ugly"

Good things about it:

- We keep relying in constraints to unique validations
- We do not need to create another validation
- No query on built-in validations
- We fix the problem which the user gets the constraint error only when other validations pass
- If the uniqueness validation change, we need to change only one place (the constraint)

nathanl commented:
@kelvinst It's an interesting idea, but constraints inherently provide incomplete feedback because they fail one at a time. Eg, if you try an `INSERT` that would violate 3 constraints, PostgreSQL will only return an error message for the first one it notices. Whereas if you checked all three conditions with validations, you'd get 3 error messages.

kelvinst commented:
> It's an interesting idea, but constraints inherently provide incomplete feedback because they fail one at a time

Uh-oh, that I was not aware of. Sorry for the bothering, nevermind what I've said. 😞

nathanl commented:
@kelvinst No worries, thanks for trying to help! :)

nathanl commented:
@josevalim Yes, the original thought was to validate uniqueness only. I made it more generic in an effort [not to couple changesets to repos or queries](https://github.com/elixir-ecto/ecto/pull/2006#issuecomment-288951980). This meant making the user pass in a "conflict finder" function, which in turn meant that it could be looking for any kind of conflict (eg overlap), not just uniqueness.

I think the point about coupling was a valid criticism, but I don't see a way to help Ecto users validate uniqueness without, at minimum, using `Ecto.Query` to generate a query for conflicting records.

If we go back to the original intent to just validate uniqueness, it could be something like this.

```elixir
  def unsafe_validate_unique(changeset, field_names, repo, error_field, error_message \\ "must be unique") do
    where_clause = Enum.map(field_names, fn (field_name) ->
      {field_name, get_field(changeset, field_name)}
    end)

    if Enum.any?(where_clause, fn (tuple) -> is_nil(elem(tuple, 1)) end) do
      changeset
    else
      dups_query = Ecto.Query.from q in __MODULE__, where: ^where_clause

      # For updates, don't flag record as a dup of itself
      changeset_id = get_field(changeset, :id)
      dups_query = if is_nil(changeset_id) do
        dups_query
      else
        from q in dups_query, where: q.id != ^changeset_id
      end

      dups_exist_query = from q in dups_query, select: true, limit: 1
      case repo.one(dups_exist_query) do
        true -> add_error(
          changeset,
          error_field,
          error_message,
          [validation: [:unsafe_validate_unique, field_names]]
        )
        nil  -> changeset
      end
    end
  end
```

If this is still too much coupling, maybe a separate Hex package would be in order?

josevalim commented:
@nathanl I would say the "coupling" is fine because the point is exactly to provide a straight-forward use case for the functionality. If we make it generic, then we go back to the point it doesn't add much compare to the existing changeset API. So I would add your function as is with two changes:

  1. Make sure to use Ecto.primary_key to get the primary keys (we can't assume ID)
  2. Assume the error_field is the first field in field_names

johnnyshields commented:
"validate_uniqueness_optimistically" or "validate_optimistic_uniqueness" might be better than "unreliably"

josevalim commented:
@nathanl do you have any plans to finish this? We are starting to work on Ecto 2.2 and it would be nice to get this in.

nathanl commented:
I'd like to, yes. The part I got stuck on was testing. I thought I would insert a record, try to validate an identical one, and see that it fails. But I couldn't see how to get the existing test repo to behave as though the first record existed.

josevalim commented:
@nathanl for the unit test, check test_repo.exs. One way to solve it is to use the process inbox or the process dictionary to instruct the adapter to return a certain value. We do so currently only on the migration tests.

For the integration test, you can just insert the record. ;)

nathanl commented:
Closing in favor of a new PR - see https://github.com/elixir-ecto/ecto/pull/2162
